docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -0,0 +1,646 @@
+        // currently a resource is being updated
+   	    ResourceUpdateInProgress bool
+   	    // Resource being updated
+   	    MemoryLimitInBytesBeingUpdated int64
+   	    ......
+   	    ......
+   	    }
+
+Add a handlers for the VIR_DOMAIN_EVENT_ID_DEVICE_REMOVED and VIR_DOMAIN_EVENT_ID_DEVICE_REMOVAL_FAILED events in Libvirt.
+
+In UpdateContainerResource(), the workflow is as follows:
+
+    1. if the VMConfig.ResourceupdteInProgress is set, ignore the call from Kubelet for VM update to the same resource 
+    2. In the AdjustDomainMemory() function, Set the VMConfig to set the ResourceUpdateInProgress
+    3. The two aforementioned handlers reset the VMConfig.ResourceupdteInProgress
+   
+   
+#### Multiple memory device sizes support
+Different flavors of VMs will likely need different level of memory scale-up to supports it peak resource consumptions. Cloud
+providers, like AWS, has VM flavors from nano type ( 1vcpu, 500MiB RAM ) to extra large or memory optimized VMs with TB level
+of RAMs. Some customer data or VM usage analysis are needed to determine what size memory device will be needed. Currently Arktos
+starts with the below sizes of devices, that supports from 128MiB to 2GiB for each device. NUMA node can be set as needed.
+
+    const memoryDeviceDefinition = `<memory model='dimm'>
+							<target>
+								<size unit='MiB'>128</size>
+								<node>0</node>
+							</target>
+						</memory>`
+
+    const memoryDeviceDefinition = `<memory model='dimm'>
+							<target>
+								<size unit='MiB'>512</size>
+								<node>0</node>
+							</target>
+						</memory>`
+
+    const memoryDeviceDefinition = `<memory model='dimm'>
+							<target>
+								<size unit='MiB'>1024</size>
+								<node>0</node>
+							</target>
+						</memory>`
+
+    const memoryDeviceDefinition = `<memory model='dimm'>
+							<target>
+								<size unit='MiB'>2048</size>
+								<node>0</node>
+							</target>
+						</memory>`
+ 
+The VM runtime internal function, determineNumberOfDeviceNeeded(),  will be modified with the workflow as:
+ 
+    If AttachDevice
+       GetVMSize
+       GetFreeSlot
+       if VMsizeSmall and FreeSlotEnough and numberOfNewDevice < "some limits"
+          DetermineNumberDevicesAndSize
+          
+    If Detach
+       QueryCurrentAttachedDevices
+       DetermineDevicesToDetach
+          
+    return ArrayOfDevieXml
+     
+ 
+#### API changes
+As described in the "CRI changes" section, the Arttos-vm-runtime UpdateContainerResource() were implemented to update
+the VM CGroup settings and update the VM domain definition with the new resource configuration.
+
+If Kubelet enables its CPU manager and set the cpus to a particular pod. the cpuset is sent to the runtime via the 
+UpdateContainerResource() and the runtime will honor the cpusets in both VM definition and its CGroup.
+
+Two new APIs, SetVcpus() and AdjustDomainMemory() were added to support VM vertical scaling, which extends the call to 
```

Comment:
Are these two APIs or just runtime internal methods? If they are not part of CRI APIs, it's better to put it under VM runtime internal implementation.

Reply:
Good point. those are not CRI APIs. i have changed this section to "Arktos-vm-runtime internal API changes" to reflect this. thanks.